### ISSUE TITLE
Daily log & To-do task

### ISSUE THREAD

**Issue by tmtmaj** [tmtmaj/security_risk_analysis_210427#1]
To-do task
-

- [ ]  catboost gridsearch
- [ ]  transformer encoder using pytorch
- [ ]  countvectorize, tdidfvectoriaze 에 대한 공부 (돌아가는 과정)

1일차
-

ML 모델에서 몇가지 시도 

- LGBM 시도

- Randomforest 시도 (baseline)

1. 0.83... (1fold)
2. 0.845... (8fold)

- Catboost 시도 

randomsearch, optuna, bayesian optimizer 모두 GPU 사용시 Featureid < cursor 라는 에러 발생 
-> catboost github에 질문 올려놓음 https://github.com/catboost/catboost/issues/1675

search에 값을 넣는 방법들

numpy.linspace

```
이 함수는 위의 numpy.arrange와 비슷하나, step 대신 배열에 생성될 맴버의 갯수로 각 값의 편차로 배열을 생성해줍니다. 

numpy.linspace(start, stop, num, endpoint, retstep, dtype)

import numpy
x = numpy.linspace(10,20, 5, endpoint = False)
print(x)

[10. 12. 14. 16. 18.]
```

numpy.logspace

```
이름 그대로 로그 스케일 값으로 배열을 만드는 함수입니다. 개인적으로 수학을 썩 좋아하지는 않지만, 은근 로그 스케일 계산해야하는 경우도 있고 합니다. 관련된 부분입니다. 
numpy.logspace(start, stop, num, endpoint, base, dtype)

import numpy
x = numpy.logspace(1,10,num = 10, base = 2) 
print(x)

[   2.    4.    8.   16.   32.   64.  128.  256.  512. 1024.]
```

- Catboost paramter tuning

--l2-leaf-reg: 

```
L2 regularization 
양수 값만 가능 
2~30이 good starting 하지만 데이터마다 다를 수 있음
```

--border_count: 

```
The number of splits for numerical features.
The default value depends on the processing unit type and other parameters:

CPU: 254
GPU in PairLogitPairwise and YetiRankPairwise modes: 32
GPU in all other modes: 128

속도를 올리기 위해서 128, 32을 사용 근데 최고의 성적을 얻어보고 싶다면 254도 써보기

```

--bootstrap_type:

```
 "Bayesian", "Bernoulli"
이것들과 매칭되는 parameter가 있음

if param["bootstrap_type"] == "Bayesian":
param["bagging_temperature"] = trial.suggest_float("bagging_temperature", 0, 10)
elif param["bootstrap_type"] == "Bernoulli":
param["subsample"] = trial.suggest_float("subsample", 0.1, 1)

```

2일차 
-

sklearn 의 grid_search로 catboost를 돌리고 있음. 근데 RAM 이 터져 버려서 
cv 수를 낮췄더니 간신히 돌아감

grid_search가 만약에 계속 memory leak이 나면 그냥 개별적으로 돌려서 확인하는 수밖에 없음.

```
from catboost import CatBoostClassifier
from sklearn.model_selection import GridSearchCV
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.feature_extraction.text import TfidfTransformer

# clf = CatBoostClassifier(task_type='GPU', border_count=None)

clf = CatBoostClassifier(
        loss_function="MultiClass",
        task_type="GPU",

        )

params = {'iterations': np.arange(100, 2000, 300),
          'depth': np.arange(3, 10, 1),
          'loss_function': ['MultiClass'],
          'l2_leaf_reg': np.arange(2,30,1),
          'learning_rate' : np.arange(0.001, 0.3, 0.003),               
        'random_strength' : np.arange(0, 100, 10),
          'random_seed': [42],
          "bootstrap_type": ["Bayesian"],
          "bagging_temperature": np.arange(0, 10, 1),
          "border_count" : [32]
         }

# params = {'iterations': np.arange(100, 2000, 100),
#           'depth': np.arange(3, 10, 1),
#           'loss_function': ['MultiClass'],
#           'l2_leaf_reg': np.arange(2,30,1),
#           'leaf_estimation_iterations': [10],
#           'learning_rate' : np.arange(0.001, 0.3, 0.003),               
#         'random_strength' : np.arange(0, 100, 10),
#           'logging_level':['Silent'],
#           'random_seed': [42],
#           "bootstrap_type": ["Bernoulli"],
#           "subsample": np.arange(0.1, 1, 0.1)
#           "border_count" : [254, 128, 32]
#          }

count_vect = CountVectorizer(analyzer="word", max_features=10000) 
train_count = count_vect.fit_transform(train_text)

tfidf_transformer = TfidfTransformer(use_idf=True, smooth_idf=True, norm='l2', sublinear_tf=True)
train_tfidf = tfidf_transformer.fit_transform(train_count).astype('float32')

from sklearn.metrics import log_loss, make_scorer, f1_score

def macro_f1(y_true, y_pred):
  y_pred = np.argmax(y_pred, -1)
  return f1_score(y_true, y_pred, average='macro')
f1 = make_scorer(macro_f1, greater_is_better=True, needs_proba=True)

clf_grid = GridSearchCV(estimator=clf, param_grid=params, scoring=f1, cv=5, verbose = 100)
clf_grid.fit(X = train_tfidf, y = train_level, verbose=200)

sorted(clf_grid.cv_results_.keys())
```

생각하는 가장 큰 문제는 Transformer의 output을 어떻게 처리할 것인가? 
(bsz, seq_length, model_dim)의 출력이 transformer로부터 나올텐데 이를 어떻게 (bsz, target_label)로 바꿔줄까? 
rnn의 경우는 하나하나 하기 때문에 마지막 output을 사용하면 되는데 transformer는 전체 seq_length를 받는 다는 것이 좀 헷갈린다. 
방법 1. 똑같이 마지막 seq_length만 뽑아서 linear(model_dim, target_label) 통과시켜서 결과를 얻는다!
방법 2.  seq_length에 대해서 평균을 낸다 그후 linear 통과
방법 3.  cls같은 토큰을 통해서 구분한다. (그냥 입력에서  [CLS] 추가하고 seq_length 0부분만 뽑아서 linear 통과시키면 되나?

- [ ] pytorch에서 제공하는 transformerencoder에 대한 입력으로 src, src_mask, src_key_padding_mask 세가지가 있다. 
src_mask는 decoder 단의 masked_self_attn을 위한 것인가? 아니면 src 측에도 넣어줘야하나?
즉, src_mask는 padding의 위치를 알려주는 건가? 아니면 autoregressive하게 만들기 위한 mask인가?
이부분 확인해 볼 필요 있다.
pytorch.org : Along with the input sequence, a square attention mask is required because the self-attention layers in nn.TransformerEncoder are only allowed to attend the earlier positions in the sequence. For the language modeling task, any tokens on the future positions should be masked. 
그렇다면 fairseq encoder에서도 이렇게 하고 있는지 확인해보자!
(사실 이렇게 되고 있다면 src측도 autoregressvie한 거라 완전 병렬이라 할 수 있나? )

**Comment by tmtmaj** [comment]
## 7일차 (2021-05-02)

- [ ] trail using transformer
- [ ] data preprocess
1. BPE
2. pruning data
